src/macaron/json_tools.py

```diff
@@ -21,8 +20,8 @@ def json_extract(entry: JsonType, keys: list[str], type_: type[T]) -> T | None:
     ----------
     entry: JsonType
         An entry point into a JSON structure.
-    keys: list[str]
-        The list of depth-sequential keys within the JSON.
+    keys: list
```

Comment:
Are we expecting elements of `keys` has any other types besides `str` and `int`. If not, I think we can use the type annotation `list[str | int]` here.

Reply:
`list[str | int]` may not be the best choice here. tldr: In this particular case, I think `Sequence[str | int]` makes the most sense.

Although `str | int` is a supertype of both `str` and `int`, intuitively we may think `list[str | int]` is also a supertype of both `list[str]` and `list[int]`. However, this is not true, because `list` in Python is invariant. For the above property to hold, `list` needs to be covariant, which it is not (See [this section](https://mypy.readthedocs.io/en/stable/generics.html#variance-of-generic-types) in the mypy doc as they explain it much better than I can). You can try the following example with mypy:

```
from typing import reveal_type

def do_it(l: list[int | str]) -> None:
    return

a = [1, 2, 3]
b = ["foo", "bar", "baz"]
c = [1, "foo"]
reveal_type(a)
reveal_type(b)
reveal_type(c)
do_it(a)  # type error
do_it(b)  # type error
do_it(c)  # no type error
```

A covariant alternative usually used in a scenario like this would be `Sequence`. If you change the type annotation of the param `l` of `do_it` to `Sequence`, there will be no type error when you call `do_it(a)` or `do_it(b)`. One caveat is that `c` needs to be type-annotated as `list[str | int]`, as `mypy` infers `c` to be a `list[object]` and raises a type error otherwise. Here is the complete example with no type errors.

```python3
from typing import Sequence, reveal_type


def do_it(l: Sequence[int | str]) -> None:
    return


a = [1, 2, 3]
b = ["foo", "bar", "baz"]
c: list[int | str] = [1, "foo"]
reveal_type(a)
reveal_type(b)
reveal_type(c)
do_it(a)
do_it(b)
do_it(c)
```

Reply:
Ah I see. Thanks for the explanation. @nathanwn suggestion sounds good to me.